Compute cached

zvookin opened the pull request:
Opening a PR to get comments. This provides zero copy memoization. I will likely work on factoring the changes to StorageFlattening into a separate pass.

steven-johnson commented:
Thought: we could reduce the overhead for calculating the hash value by precalculating the 'fixed' portion (toplevel name + key name), either at compile time (if we're willing to bake assumptions about the hash function) or once-at-runtime (on first call). Patching this in and trying it, I get cache keys of ~72 bytes, of which only the last ~8 bytes vary with input parameters. (Admittedly, saving the hash calculation of 64 bytes may be a grasping at optimizations...)

zvookin commented:
This is ready to review. I think there may be one docs TODO, which I'll check on tomorrow. Also it would be great if others could look at the USE_FULL_NAMES_IN_KEY #ifdef in Memoization.cpp. It addresses the comment above from Steven by using a pointer to a constant string and a 32-bit counter to identify the filter and Func in the cache key instead of each of their names as strings (with lengths as well).

The PR should make Memoization much faster. It is techncially a breaking change in that one can no longer memoize an output. That now results in a compile time error. The fix is to insert a compute_root'ed Func before the output Func and memoize the inner Func instead. (There is a change to the memoize.cpp test which demonstrates this.)

There is some question as to whether separating out the rewrite_memoized_allocations pass from storage_flattening is a good idea. The rewrite ends up being very dependent on the IR that storage flattening generates.

steven-johnson commented:
Something is wonky with the latest changes; I pulled into a local test and it seems to get caught in an endless loop. Investigating.

steven-johnson commented:
Edit: it's not endless, but broken: with USE_FULL_NAMES_IN_KEY=1 my cache hit rate is very close to 100%; with USE_FULL_NAMES_IN_KEY=0 it's under 30%.

steven-johnson commented:
Looks like this line:
         size_so_far = (size_so_far + needed_alignment) & ~(needed_alignment - 1);
should be:
         size_so_far = (size_so_far + needed_alignment - 1) & ~(needed_alignment - 1);

(As it is, we round up the allocation size more than we need and end up with uninitialized bytes in the cache key.)

zvookin commented:
Fix pushed.

On Mon, Jul 27, 2015 at 11:27 AM, Steven Johnson <notifications@github.com>
wrote:

> Looks like this line:
> size_so_far = (size_so_far + needed_alignment) & ~(needed_alignment - 1);
> should be:
> size_so_far = (size_so_far + needed_alignment - 1) & ~(needed_alignment -
> 1);
>
> (As it is, we round up the allocation size more than we need and end up
> with uninitialized bytes in the cache key.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/871#issuecomment-125296882>.
>

steven-johnson commented on src/Memoization.cpp:
```diff
@@ -162,6 +163,17 @@ class KeyInfo {
+// if a pointer is used, a counter must also be put in the cache key
+// to avoid aliasing on reuse of the address in JIT situations where
+// code is regenerated into the same region of memory.
+#define USE_FULL_NAMES_IN_KEY 0
```
Obviously we need to decide on this before landing. On my pathological-ish test case it definitely moved the needle performance wise so I vote for setting it to zero.

steven-johnson commented on src/runtime/cache.cpp:
```diff
@@ -343,15 +369,28 @@ WEAK bool halide_memoization_cache_lookup(void *user_context, const uint8_t *cac
                 for (int32_t i = 0; i < tuple_count; i++) {
                     buffer_t *buf = tuple_buffers[i];
-                    copy_from_to(user_context, entry->buffer(i), *buf);
+                    //copy_from_to(user_context, entry->buffer(i), *buf);
```
Why is this commented out? Explain or remove, methinks

steven-johnson commented on src/runtime/cache.cpp:
```diff
@@ -343,15 +369,28 @@ WEAK bool halide_memoization_cache_lookup(void *user_context, const uint8_t *cac
+    for (int32_t i = 0; i < tuple_count; i++) {
+        buffer_t *buf = tuple_buffers[i];
+        size_t buffer_size = full_extent(*buf);
+        // TODO: ERROR RETURN
```
Is this TODO still a thing? If so I think it needs some fleshing out

zvookin commented:
I fixed the TODOs by adding memory allocation failure logic and a test to make sure it works. Only remaining thing I know of is to decide what to do about the USE_FULL_NAMES_IN_KEY ifdefs.

steven-johnson commented:
> what to do about the USE_FULL_NAMES_IN_KEY

Normally I'm reluctant to leave ifdef'ed code in place, but since there's a measurable performance win for the =0 case and a meaningful debugging win for the =1 case, tempted to leave it there, defaulted to zero.

zvookin commented:
I updated the USE_FULL_NAMES_IN_KEY comment per Steven's last reply. Barring more review feedback or test failures in broader testing, I think this is ready to go.

abadams commented on src/IR.h:
```diff
@@ -271,17 +271,25 @@ struct Allocate : public StmtNode<Allocate> {
+
+    // These override the code generator dependent malloc and freeequivalents
+    // if provided. If the new succeeds, the delete is guaranteed to be called.
+    // It is allowed for the delete_stmt to be undefined even if
```
delete_stmt no longer exists

abadams commented on src/IR.h:
```diff
@@ -271,17 +271,25 @@ struct Allocate : public StmtNode<Allocate> {
     std::vector<Expr> extents;
     Expr condition;
+
+    // These override the code generator dependent malloc and freeequivalents
```
freeequivalents

steven-johnson replied:
NEW WORD OF THE DAY

abadams commented on src/runtime/cache.cpp:
```diff
@@ -70,20 +70,16 @@ WEAK size_t full_extent(const buffer_t &buf) {
 }
 
 WEAK void copy_from_to(void *user_context, const buffer_t &from, buffer_t &to) {
```
This function now seems misleadingly named.

abadams commented on src/runtime/cache.cpp:
```diff
@@ -70,20 +70,16 @@ WEAK size_t full_extent(const buffer_t &buf) {
 }
 
 WEAK buffer_t copy_of_buffer(void *user_context, const buffer_t &buf) {
     buffer_t result = buf;
```
Is this function a noop? It looks like it just returns a shallow copy of the buffer_t now.

abadams commented on src/runtime/windows_clock.cpp:
```diff
@@ -35,7 +35,7 @@ WEAK int64_t halide_current_time_ns(void *user_context) {
 
 WEAK void halide_sleep_ms(void *user_context, int ms) {
-	Sleep(ms);
+        Sleep(ms);
```
Looks like I introduced these tabs via msvc, and this PR changes them to 8 spaces. Mind doing a drive-by conversion to 4 spaces in this PR? (to avoid the merge conflict if I fix in master)

abadams commented:
LGTM. Will test on Windows and report back.

abadams commented:
Crashes randomly on win32. Turning on using full names in key seems to fix it.

abadams commented:
Good news! Also crashes randomly on linux in 32-bit mode. That should make the debugging easier.

steven-johnson commented:
Hmm, I think Handle().bytes() is always 64 bits, even for 32-bit mode... in which case we could have uninitialized bytes in the upper (lower?) half.

abadams commented:
Yeah, that looked like the issue. The fix was non-trivial, so I made it a PR into master ( #880 )
